library/fluid/fluid_extensions/src/main/java/org/quiltmc/qsl/fluid/api/FlowableFluidExtensions.java

```diff
@@ -0,0 +1,372 @@
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.loot.LootTables;
+import net.minecraft.particle.ParticleEffect;
+import net.minecraft.particle.ParticleTypes;
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.sound.SoundEvents;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.random.RandomGenerator;
+import net.minecraft.world.World;
+import net.minecraft.world.event.GameEvent;
+
+import javax.annotation.Nullable;
+
+public interface FlowableFluidExtensions {
+
+	float WATER_VISCOSITY = 0.8f;
+	float LAVA_VISCOSITY = 0.5f;
+	float WATER_PUSH_STRENGTH = 0.014f;
+	float LAVA_PUSH_STRENGTH_OVERWORLD = 7 / 3000f;
+	float LAVA_PUSH_STRENGTH_ULTRAWARM = 0.007f;
+	float WATER_DENSITY = 1000f;
+	float LAVA_DENSITY = 3100f;
+	float WATER_TEMPERATURE = 300f;
+	float LAVA_TEMPERATURE = 1500f;
+	float WATER_FOG_START = -8.0f;
+	float FULL_FALL_DAMAGE_REDUCTION = 0f;
+	float HALF_FALL_DAMAGE_REDUCTION = 0.5f;
+	float NO_FALL_DAMAGE_REDUCTION = 0f;
+	int WATER_FOG_COLOR = -1;
+
+	Identifier WATER_FISHING_LOOT_TABLE = LootTables.FISHING_GAMEPLAY;
+
+	/**
+	 * The color of this fluid.
+	 */
+	default int getColor(FluidState state, World world, BlockPos pos) {
+		return world.getBiome(pos).value().getWaterColor();
+	}
+
+	/**
+	 * 0.8F is the default for water
+	 * 0.5F is the default for lava
+	 *
+	 * @see FlowableFluidExtensions#WATER_VISCOSITY
+	 * @see FlowableFluidExtensions#LAVA_VISCOSITY
+	 */
+	default float getHorizontalViscosity(FluidState state, Entity affected) {
+		return WATER_VISCOSITY;
+	}
+
+	/**
+	 * Default for water and lava is 0.8F
+	 *
+	 * @see FlowableFluidExtensions#WATER_VISCOSITY
+	 */
+	default float getVerticalViscosity(FluidState state, Entity affected) {
+		return WATER_VISCOSITY;
+	}
+
+	/**
+	 * 0.014F is the default for water
+	 * 7 / 3000 is the default for lava in the overworld
+	 * 0.007F is the default for lava in the nether
+	 *
+	 * @see FlowableFluidExtensions#WATER_PUSH_STRENGTH
+	 * @see FlowableFluidExtensions#LAVA_PUSH_STRENGTH_OVERWORLD
+	 * @see FlowableFluidExtensions#LAVA_PUSH_STRENGTH_ULTRAWARM
+	 */
+	default float getPushStrength(FluidState state, Entity affected) {
+		return WATER_PUSH_STRENGTH;
+	}
+
+	/**
+	 * Toggles weather or not a player can sprint swim in your fluid
+	 */
+	default boolean canSprintSwim(FluidState state, Entity affected) {
+		return true;
+	}
+
+	/**
+	 * @return an updated horizontalViscosity, specifically regarding potion effects
+	 */
+	default float modifyHorizontalViscosity(LivingEntity affected, float horizontalViscosity) {
+		if (affected.hasStatusEffect(StatusEffects.DOLPHINS_GRACE)) {
+			horizontalViscosity = 0.96f;
+		}
+		return horizontalViscosity;
+	}
+
+	default boolean enableSpacebarSwimming(FluidState state, Entity affected) {
```

Comment:
```suggestion
	default boolean allowSpacebarSwimming(FluidState state, Entity affected) {
```
I also have no idea what "spacebar swimming" means. This either needs Javadoc or a better name.